### ISSUE TITLE
Unable to run Julia, terminal process... failed to launch (exit code: 1)

### ISSUE THREAD

**Issue by Kevin-Mattheus-Moerman** [julia-vscode/julia-vscode#2491]
I have just switched from Juno to VSCode so quite new at this. After a fresh install of VSCode I tried to run a script but keep getting this error: 

`The terminal process "julia '-i', '--banner=no', '--project=/home/kevin/.julia/environments/v1.6', '/home/kevin/.vscode/extensions/julialang.language-julia-1.4.3/scripts/terminalserver/terminalserver.jl', '/tmp/vsc-jl-repl-6b958567-c4b2-4e38-a5ac-fb792682717b', '/tmp/vsc-jl-cr-adc76cdd-4c04-4192-bf90-b272f2f92fd8', 'USE_REVISE=true', 'USE_PLOTPANE=true', 'USE_PROGRESS=true', 'DEBUG_MODE=false'" failed to launch (exit code: 1).`

The gif below shows that I can run a python script but I cannot run Julia.

Perhaps it relates to this as I very briefly see `execvp(3) failed.: Permission denied`:
https://github.com/microsoft/vscode/issues/124968

Can you help trouble shoot this? Thanks. 

Details:
Ubuntu Ubuntu 20.04.3 LTS
Julia Version 1.6.2 (2021-07-14)

![Peek 2021-10-05 09-38](https://user-images.githubusercontent.com/8392709/135989923-560daed2-e19d-477c-80c1-449d448567f6.gif)

**Comment by Kevin-Mattheus-Moerman** [comment]
Not sure if this helps (or if I'm supposed to try this) but I tried that command in a terminal window and got:

<html>
<pre>julia -i --banner=no --project=/home/kevin/.julia/environments/v1.6 /home/kevin/.vscode/extensions/julialang.language-julia-1.4.3/scripts/terminalserver/terminalserver.jl /tmp/vsc-jl-repl-6b958567-c4b2-4e38-a5ac-fb792682717b  /tmp/vsc-jl-cr-adc76cdd-4c04-4192-bf90-b272f2f92fd8  USE_REVISE=true  USE_PLOTPANE=true  USE_PROGRESS=true  DEBUG_MODE=false
<font color="#EF2929"><b>ERROR: </b></font>LoadError: IOError: connect: connection refused (ECONNREFUSED)
Stacktrace:
 [1] <b>wait_connected(</b><font color="#555753">x</font>::Base.PipeEndpoint<b>)</b>
<font color="#555753">   @ </font><font color="#75507B">Sockets</font> <font color="#555753">/buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/Sockets/src/</font><font color="#555753"><u style="text-decoration-style:single">Sockets.jl:532</u></font>
 [2] <b>connect</b>
<font color="#555753">   @ /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/Sockets/src/</font><font color="#555753"><u style="text-decoration-style:single">Sockets.jl:567</u></font><font color="#555753"> [inlined]</font>
 [3] <b>connect</b>
<font color="#555753">   @ /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/Sockets/src/</font><font color="#555753"><u style="text-decoration-style:single">PipeServer.jl:97</u></font><font color="#555753"> [inlined]</font>
 [4] <b>serve(</b><font color="#555753">args</font>::String; <font color="#555753">is_dev</font>::Bool, <font color="#555753">crashreporting_pipename</font>::String<b>)</b>
<font color="#555753">   @ </font><font color="#06989A">VSCodeServer</font> <font color="#555753">~/.vscode/extensions/julialang.language-julia-1.4.3/scripts/packages/VSCodeServer/src/</font><font color="#555753"><u style="text-decoration-style:single">VSCodeServer.jl:99</u></font>
 [5] top-level scope
<font color="#555753">   @ ~/.vscode/extensions/julialang.language-julia-1.4.3/scripts/terminalserver/</font><font color="#555753"><u style="text-decoration-style:single">terminalserver.jl:45</u></font>
in expression starting at /home/kevin/.vscode/extensions/julialang.language-julia-1.4.3/scripts/terminalserver/terminalserver.jl:24
</pre>

**Comment by pfitzseb** [comment]
Can you try the `Julia: Connect external REPL` command? That will prompt you to copy-paste a code snippet into an already running Julia session and might give a more informative error message (the `ECONNREFUSED` probably happens because VS Code isn't creating a pipe for Julia to connect to).

**Comment by Kevin-Mattheus-Moerman** [comment]
@pfitzseb thanks for your help. 

Where/how do I issue this command?

**Comment by pfitzseb** [comment]
Hit Ctrl-Shift-P in VS Code to bring up the command palette.

**Comment by Kevin-Mattheus-Moerman** [comment]
It produced: 
`pushfirst!(LOAD_PATH, raw"/home/kevin/.vscode/extensions/julialang.language-julia-1.4.3/scripts/packages");using VSCodeServer;popfirst!(LOAD_PATH);VSCodeServer.serve(raw"/tmp/vsc-jl-repl-246f75da-a1f1-4917-b546-616271874232"; is_dev = "DEBUG_MODE=true" in Base.ARGS, crashreporting_pipename = raw"/tmp/vsc-jl-cr-86134bf0-0c72-4ea5-9833-94cdf66dd612");nothing # re-establishing connection with VSCode`

I typed `julia` in the terminal and pasted the above. It looks like it worked the first time to trigger julia to start. However when I closed and reopened VSStudio the above problem persistent. I then retried pasting this snippet which then produced: 

```
ERROR: IOError: connect: no such file or directory (ENOENT)
Stacktrace:
 [1] wait_connected(x::Base.PipeEndpoint)
   @ Sockets /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/Sockets/src/Sockets.jl:532
 [2] connect
   @ /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/Sockets/src/Sockets.jl:567 [inlined]
 [3] connect
   @ /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.6/Sockets/src/PipeServer.jl:97 [inlined]
 [4] serve(args::String; is_dev::Bool, crashreporting_pipename::String)
   @ VSCodeServer ~/.vscode/extensions/julialang.language-julia-1.4.3/scripts/packages/VSCodeServer/src/VSCodeServer.jl:99
 [5] top-level scope
   @ REPL[1]:1
```

**Comment by pfitzseb** [comment]
That's wild. There should be absolutely no difference between doing that and what we automatically do for you when you try to run a file.

Just to be sure: `Julia: Start REPL` produces the same error as in the OP? Also, can you try temporarily disabling the Python extension?

**Comment by Kevin-Mattheus-Moerman** [comment]
@pfitzseb yes that produces the same error and I have the python extension disabled (and restarted VSStudio).

**Comment by Kevin-Mattheus-Moerman** [comment]
Is it correct that the code snippet is different every time? e.g. first time: 
`pushfirst!(LOAD_PATH, raw"/home/kevin/.vscode/extensions/julialang.language-julia-1.4.3/scripts/packages");using VSCodeServer;popfirst!(LOAD_PATH);VSCodeServer.serve(raw"/tmp/vsc-jl-repl-4cb8f992-435a-4d74-906b-398000575fc6"; is_dev = "DEBUG_MODE=true" in Base.ARGS, crashreporting_pipename = raw"/tmp/vsc-jl-cr-32634856-bb83-4a1a-a9a8-b7b4f0cf7021");nothing # re-establishing connection with VSCode`

Second time:
`pushfirst!(LOAD_PATH, raw"/home/kevin/.vscode/extensions/julialang.language-julia-1.4.3/scripts/packages");using VSCodeServer;popfirst!(LOAD_PATH);VSCodeServer.serve(raw"/tmp/vsc-jl-repl-dd8e288e-7f65-4646-b031-e782dd816405"; is_dev = "DEBUG_MODE=true" in Base.ARGS, crashreporting_pipename = raw"/tmp/vsc-jl-cr-b49629d4-2ee4-4629-8997-a933060ef158");nothing # re-establishing connection with VSCode`

**Comment by pfitzseb** [comment]
Yeah, that's expected. We open a new pipe every time you use that command.

**Comment by Kevin-Mattheus-Moerman** [comment]
Good news, I can get VS Code to call julia properly now if I manually specify the path in the Julia VS Code extension (which was empty), 

![Screenshot from 2021-10-12 19-31-34](https://user-images.githubusercontent.com/8392709/137010188-af17590c-48e0-409c-8e37-16de13d8d22e.png)

**Comment by pfitzseb** [comment]
We just use "julia" (and check a few paths on OSs with default install locations) if that setting is empty.

Anyways, I'm glad you got the extension up and running.

**Comment by graidl** [comment]
I ran into the exactly same issue as described above, although with an additional layer of complexity: I use a remote ssh connection / remote vscode server.

To summarize:

- Starting the remote vscode session works perfectly well.
- "Run file in new process" works.
- I can also open new (remote) bash-terminals in vscode.
- Also, starting julia in such a terminal works.
- "Connect external REPL" works with this terminal as expected, by pasting the provided code in the terminal
- However, a "Start REPL" instead of the above fails with "The terminal process...terminated with exit code: 1"
- The main problem now is that there is no "Julia: Executable Path" setting when working with a remote connection, therefore the above solution doesn't work in this setting.
- Julia is locally installed in my home directory and ~/bin/julia links to it, ~/bin is in PATH
Any idea how I can resolve this problem?

**Comment by pfitzseb** [comment]
> The main problem now is that there is no "Julia: Executable Path" setting when working with a remote connection, therefore the above solution doesn't work in this setting.

That setting exists for sure. Make sure to switch to the remote tab at the top (and that you have the Julia extension installed on the remote):
![image](https://user-images.githubusercontent.com/6735977/143549125-bf913390-c230-4f82-8f39-f9e0beffe96d.png)

**Comment by graidl** [comment]
Thanks for the quick response, indeed I have completely missed this remote settings tab!
Unfortunately, setting this executable path still does not resolve the issue in my case, I still get "The terminal process...terminated with exit code: 1", while "Connect external REPL" works.

Any idea how I can find out more about why the terminal process terminated?

**Comment by pfitzseb** [comment]
Can you try the `Run File in New Process` command? That should keep the terminal open after Julia dies.

**Comment by graidl** [comment]
`Run File in New Process` works without issues: Julia runs my program and then terminates, with the terminal kept open to be reused.

**Comment by pfitzseb** [comment]
Hm, interesting. What happens if you enable `julia.persistentSession.enabled` and then try to start a new REPL session?

**Comment by graidl** [comment]
This helped! I could now see an error message saying that a certain symbol was not found in the `JuliaSysimage.so` of my environment, and indeed this system image was an outdated one from a former Julia version. After removing it, starting the REPL works fine.

Thanks a lot for your much appreciated help! 

This, however, also implies that system images are so far not used in `Run File in New Process`, right? Just to understand why this worked before.

**Comment by pfitzseb** [comment]
Yes, looks like it. Did you specify the sysimg yourself in the `julia.additionalArgs` setting or is that one the you generated in VS Code that's picked up automatically?

**Comment by graidl** [comment]
I didn't specify it anywhere explicitly, my `julia.additionalArgs` is empty.
